File: owasp4net.injection.data/NorthwindRepositoryUnsafe.vb
Imports System.Data.SqlClient
Imports owasp4net.injection.data

Public Class NorthWindRepositoryUnsafe
    Implements INorthWindRepository

    Private _connectionString As String

    Public Sub New(connectionString As String)
        If String.IsNullOrWhiteSpace(connectionString) Then
            Throw New ArgumentNullException("connectionString")
        End If

        Me._connectionString = connectionString
    End Sub

    Public Function LoadProducts(CategoryId As String) As IEnumerable(Of Product) Implements INorthWindRepository.LoadProducts

        Dim products As New List(Of Product)

        Dim query As String = "SELECT * FROM Products"

        If Not String.IsNullOrWhiteSpace(CategoryId) Then
            query &= " WHERE CategoryId = " & CategoryId
        End If

        Using connection = New SqlConnection(Me._connectionString)
            Dim command As New SqlCommand(query, connection)
            command.Connection.Open()
            Dim reader = command.ExecuteReader()

            While reader.Read()
                products.Add(New Product() With {
                    .ProductID = CInt(reader(0)),
                    .ProductName = CStr(If(IsDBNull(reader(1)), Nothing, reader(1))),
                    .SupplierID = CInt(If(IsDBNull(reader(2)), -1, reader(2))),
                    .CategoryID = CInt(If(IsDBNull(reader(3)), -1, reader(3))),
                    .QuantityPerUnit = CStr(If(IsDBNull(reader(4)), Nothing, reader(4))),
                    .UnitPrice = CDec(If(IsDBNull(reader(5)), -1, reader(5))),
                    .UnitsInStock = CInt(If(IsDBNull(reader(6)), -1, reader(6))),
                    .UnitsOnOrder = CInt(If(IsDBNull(reader(7)), -1, reader(7))),
                    .ReorderLevel = CInt(If(IsDBNull(reader(8)), -1, reader(8))),
                    .Discontinued = CBool(If(IsDBNull(reader(9)), False, reader(9)))
                })
            End While
        End Using

        Return products
    End Function

    Public Function LoadCustomers() As IEnumerable(Of Customer) Implements INorthWindRepository.LoadCustomers
        Throw New NotImplementedException()
    End Function

    Public Function LoadCustomerById(customerId As String) As Customer Implements INorthWindRepository.LoadCustomerById
        Throw New NotImplementedException()
    End Function
End Class

File: owasp4net.injection.data/NorthWindRepositoryUnsafe.cs
using Microsoft.Data.SqlClient;
using owasp4net.injection.domain;
using System;
using System.Collections.Generic;

namespace owasp4net.injection.data
{
    public class NorthWindRepositoryUnsafe : INorthWindRepository
    {
        private string _connectionString;

        public NorthWindRepositoryUnsafe(string connectionString)
        {
            if (string.IsNullOrWhiteSpace(connectionString))
                throw new ArgumentNullException("connectionString");

            this._connectionString = connectionString;
        }

        public IEnumerable<Product> LoadProducts(string CategoryId)
        {
            List<Product> products = new List<Product>();

            string query = "SELECT * FROM Products";

            if (!string.IsNullOrWhiteSpace(CategoryId))
                query += " WHERE CategoryId = " + CategoryId;

            using (var connection = new SqlConnection(this._connectionString))
            {
                SqlCommand command = new SqlCommand(query, connection);
                command.Connection.Open();
                var reader = command.ExecuteReader();

                while (reader.Read())
                    products.Add(new Product()
                    {
                        ProductID = System.Convert.ToInt32(reader[0]),
                        ProductName = System.Convert.ToString(System.Convert.IsDBNull(reader[1]) ? null : reader[1]),
                        SupplierID = System.Convert.ToInt32(System.Convert.IsDBNull(reader[2]) ? -1 : reader[2]),
                        CategoryID = System.Convert.ToInt32(System.Convert.IsDBNull(reader[3]) ? -1 : reader[3]),
                        QuantityPerUnit = System.Convert.ToString(System.Convert.IsDBNull(reader[4]) ? null : reader[4]),
                        UnitPrice = System.Convert.ToDecimal(System.Convert.IsDBNull(reader[5]) ? -1 : reader[5]),
                        UnitsInStock = System.Convert.ToInt32(System.Convert.IsDBNull(reader[6]) ? -1 : reader[6]),
                        UnitsOnOrder = System.Convert.ToInt32(System.Convert.IsDBNull(reader[7]) ? -1 : reader[7]),
                        ReorderLevel = System.Convert.ToInt32(System.Convert.IsDBNull(reader[8]) ? -1 : reader[8]),
                        Discontinued = System.Convert.ToBoolean(System.Convert.IsDBNull(reader[9]) ? false : reader[9])
                    });
            }

            return products;
        }

        public IEnumerable<Customer> LoadCustomers()
        {
            throw new NotImplementedException();
        }

        public Customer LoadCustomerById(string customerId)
        {
            throw new NotImplementedException();
        }

        public void SaveCustomer(Customer customer)
        {
            throw new NotImplementedException();
        }
    }
}